### ISSUE TITLE
KPL format aggregated logs

### ISSUE THREAD

**Issue by wryun** [awslabs/aws-fluent-plugin-kinesis#43]
Here is a quick hack of this project to do KPL format aggregated logs (wrote it as a drastic change, because format/emit end up being different again, there are already a lot of ifs for not much code, and aggregation was semi-rejected in #24 - nevertheless, maybe you want this as part of this library?)

https://github.com/atlassian/fluent-plugin-kinesis-aggregation

Reason:
https://github.com/atlassian/fluent-plugin-kinesis-aggregation#before-you-start

The important part of the code:
https://github.com/atlassian/fluent-plugin-kinesis-aggregation/blob/master/lib/fluent/plugin/out_kinesis-aggregation.rb#L107

**Comment by riywo** [comment]
Hi @wryun ,

This is great. I'm gonna start to think about the next major version to support KPL format.

**Comment by wryun** [comment]
Cool! The tricky thing about how I've implemented it is that - unlike a proper KPL producer - it doesn't support multiple partition keys. This suits our use case, but I'm not sure how appropriate it is for a general library, which may need the full KPL functionality.

**Comment by wryun** [comment]
Further testing revealed that the use of non-google protobuf libraries caused excessive memory usage.

I've updated it to use google-protobuf, but that has a couple of unfortunate caveats. So, even further from upstreaming, I suspect :(

**Comment by riywo** [comment]
Hi @wryun ,

I pushed v1.0 branch and it fully integrates with KPL binary. Could you try it if you have time?

**Comment by wryun** [comment]
Ah, excellent! Really happy that you've found the time to work on this (and the KPL wrapper - is this going to be a separate lib?), though the code looks much more complex now :)

I'm probably not going to touch our logging in the near future, since it's finally working smoothly after we switched to use the aggregation plugin. I am curious about the resource usage of this, since it's something we've struggled with. fluentd's low memory profile tended to go pear-shaped for us with the combination of the AWS SDK and ruby-based protobuf libraries (easily hitting 250mb+ under even moderate load with 4 threads; switching to the google protobuf library reduced to around 140mb). Though it may be faster, by design it looks like this would use significantly more CPU, and possibly more memory (depending on how much you gain by pushing some of the work to the KPL).
